### PR TITLE
Add XML and YAML support for new Resource attribute

### PR DESCRIPTION
**Pull request by vincentchalamon**
| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | https://github.com/api-platform/core/projects/10#card-66986300
| License       | MIT
| Doc PR        | https://github.com/api-platform/docs/pull/1473

- [x] resource/operations: defaults, requirements, options, status, schemes
- [x] properties: defaults, example, builtinTypes, schema
- [x] update doc (xml/yaml)
- [x] test that if a new attribute is added on ApiResource class, it must be added on XML (XmlAdapter)
- [x] test that if a new attribute is added on ApiResource class, it must be added on YAML (YamlAdapter)
- [x] fix tests

### CODE REVIEW

File: src/Metadata/Extractor/XmlExtractor.php [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            $resourceClass = $this->resolve((string) $resource['class']);
+
+            $this->resources[$resourceClass][] = [
+                'shortName' => $this->phpize($resource, 'shortName', 'string'),
+                'description' => $this->phpize($resource, 'description', 'string'),
+                'uriTemplate' => $this->phpize($resource, 'uriTemplate', 'string'),
+                'types' => $resource['types'] ?? null,
+                'operations' => $this->getOperations($resource, 'operation'),
+                'graphql' => $this->getOperations($resource, 'operation', true),
+                'attributes' => $this->getAttributes($resource, 'attribute') ?: null,
```
Comment (soyuka) [inline src/Metadata/Extractor/XmlExtractor.php:53]: attributes got renamed extraProperties

File: src/Metadata/Extractor/XmlExtractor.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        foreach ($resource->property as $property) {
+            $properties[(string) $property['name']] = [
+                'description' => $this->phpize($property, 'description', 'string'),
+                'readable' => $this->phpize($property, 'readable', 'bool'),
+                'writable' => $this->phpize($property, 'writable', 'bool'),
+                'readableLink' => $this->phpize($property, 'readableLink', 'bool'),
+                'writableLink' => $this->phpize($property, 'writableLink', 'bool'),
+                'required' => $this->phpize($property, 'required', 'bool'),
+                'identifier' => $this->phpize($property, 'identifier', 'bool'),
+                'iri' => $this->phpize($property, 'iri', 'string'),
```
Comment (soyuka) [inline src/Metadata/Extractor/XmlExtractor.php:122]: iri got renamed `types` and is now an array

File: src/Metadata/Extractor/XmlExtractor.php [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                'attributes' => $this->getAttributes($property, 'attribute'),
+                'subresource' => $property->subresource ? [
```
Comment (soyuka) [inline src/Metadata/Extractor/XmlExtractor.php:124]: subresource should not exist anymore

File: src/Metadata/Extractor/YamlExtractor.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                'description' => $this->phpize($propertyValues, 'description', 'string'),
+                'readable' => $this->phpize($propertyValues, 'readable', 'bool'),
+                'writable' => $this->phpize($propertyValues, 'writable', 'bool'),
+                'readableLink' => $this->phpize($propertyValues, 'readableLink', 'bool'),
+                'writableLink' => $this->phpize($propertyValues, 'writableLink', 'bool'),
+                'required' => $this->phpize($propertyValues, 'required', 'bool'),
+                'identifier' => $this->phpize($propertyValues, 'identifier', 'bool'),
+                'iri' => $this->phpize($propertyValues, 'iri', 'string'),
+                'attributes' => $propertyValues['attributes'] ?? [],
+                'subresource' => $propertyValues['subresource'] ?? null,
```
Comment (soyuka) [inline src/Metadata/Extractor/YamlExtractor.php:127]: same comments as above

File: src/Metadata/Resource/Factory/ExtractorResourceMetadataCollectionFactory.php [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+     */
+    public function create(string $resourceClass): ResourceMetadataCollection
+    {
+        $resourceMetadataCollection = new ResourceMetadataCollection($resourceClass);
+        if ($this->decorated) {
+            $resourceMetadataCollection = $this->decorated->create($resourceClass);
+        }
+
+        if (Comment::class === $resourceClass) {
+            dd(class_exists($resourceClass), $this->extractor->getResources());
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/ExtractorResourceMetadataCollectionFactory.php:60]: dd? :D

File: src/Metadata/Resource/Factory/ExtractorResourceMetadataCollectionFactory.php [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        foreach (get_class_methods($resource) as $methodName) {
+            if (0 !== strpos($methodName, 'get')) {
+                continue;
+            }
+
+            if (!method_exists($operation, $methodName) || $operation->{$methodName}()) {
+                continue;
+            }
+
+            if (!$value = $resource->{$methodName}()) {
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/ExtractorResourceMetadataCollectionFactory.php:161]: `[]` `false` and `0` may be considered as valid values use `null ===`

File: src/Metadata/Resource/Factory/ExtractorResourceMetadataCollectionFactory.php (hunk 5 above)
Comment (vincentchalamon) [reply src/Metadata/Resource/Factory/ExtractorResourceMetadataCollectionFactory.php:60]: Work in progress :wink: 
The Behat tests and the console never go through this point, even if the Comment entity is enabled

File: src/Metadata/Extractor/AbstractPropertyExtractor.php [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+                return (string) $resolved;
+            }
+
+            throw new \RuntimeException(sprintf('The container parameter "%s", used in the resource configuration value "%s", must be a string or numeric, but it is of type %s.', $parameter, $value, \gettype($resolved)));
+        }, $value);
+
+        return str_replace('%%', '%', $escapedValue);
+    }
+}
```
Comment (soyuka) [inline src/Metadata/Extractor/AbstractPropertyExtractor.php:127]: why is this needed ?

File: src/Metadata/Extractor/XmlResourceExtractor.php [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+
+        $uriVariables = [];
+        foreach ($resource->uriVariables->uriVariable as $data) {
+            $parameterName = (string) $data['parameterName'];
+            if (1 === \count($data->attributes())) {
+                $uriVariables[$parameterName] = $parameterName;
+                continue;
+            }
+
+            if (null !== ($property = $this->phpize($data, 'fromProperty', 'string'))) {
```
Comment (soyuka) [inline src/Metadata/Extractor/XmlResourceExtractor.php:164]: does this needs to be `null !==` or can it be `!($property = $this->phpize($data, 'fromProperty', 'string'))` ?

File: src/Metadata/Extractor/XmlPropertyExtractor.php [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                continue;
+            }
+
+            $data[] = isset($value->values) ? $this->buildValues($value->values) : (string) $value;
+        }
+
+        return $data;
+    }
+
+    private function buildArrayValue(?\SimpleXMLElement $resource, string $key, $default = null)
```
Comment (soyuka) [inline src/Metadata/Extractor/XmlPropertyExtractor.php:107]: This method, as well as `phpize` and maybe others are used into XmlPropertyExtractor and XmlResourceExtractor should we use a trait with these to reduce the code base?

File: src/Metadata/Extractor/YamlResourceExtractor.php [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+            'normalizationContext' => $this->buildAttribute($resource, 'normalizationContext'),
+            'denormalizationContext' => $this->buildAttribute($resource, 'denormalizationContext'),
+            'validationContext' => $this->buildAttribute($resource, 'validationContext'),
+            'filters' => $this->buildAttribute($resource, 'filters'),
+            'order' => $this->buildAttribute($resource, 'order'),
+            'extraProperties' => $this->buildAttribute($resource, 'extraProperties'),
+            'mercure' => $this->buildMercure($resource),
+            'messenger' => $this->buildMessenger($resource),
+        ];
+    }
```
Comment (soyuka) [inline src/Metadata/Extractor/YamlResourceExtractor.php:156]: Do you think that we could use the `Metadata/ApiResource` class to build these? I'm thinking of something using reflection:

```
      $refl = new \ReflectionClass(ApiResource::class);
      foreach($refl->getMethods() as $method) {
        if(0 !== strpos($method->getName(), 'get')) {continue;}
        $this->phpize($resource, strtolower(str_replace('get', $method->getName())), $method->getReturnType()); //also replace `?` in the return type
      }
```

we'd probably need to check filters, mercure and messenger. Then `buildAttributes` could be inside `phpize` when the type is an array.

**Review by soyuka — COMMENTED** [review]
huge work! Few comments.

File: src/Metadata/Extractor/AbstractPropertyExtractor.php (hunk 7 above)
Comment (vincentchalamon) [reply src/Metadata/Extractor/AbstractPropertyExtractor.php:127]: It's from the original [AbstractExtractor](https://github.com/api-platform/core/blob/main/src/Core/Metadata/Extractor/AbstractExtractor.php)

File: src/Metadata/Extractor/YamlResourceExtractor.php (hunk 10 above)
Comment (vincentchalamon) [reply src/Metadata/Extractor/YamlResourceExtractor.php:156]: I will still need to filter which property is in the `buildBase` method or not. Those properties can be declared at the resource level, but also overriden by operation, and not all properties can be declared on both sides. That's why I had to filter them here

**Review by soyuka — APPROVED** [review]
